<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Genblast Module Documentation &#8212; ensembl-anno 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/agogo.css?v=0438b505" />
    <script src="_static/documentation_options.js?v=2709fde1"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="canonical" href="https://ensembl.github.io/ensembl-anno/genblast.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Minimap2 Module Documentation" href="minimap.html" />
    <link rel="prev" title="Eponine Module Documentation" href="eponine.html" /> 
  </head><body>
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="index.html">ensembl-anno 0.1 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="eponine.html" title="Eponine Module Documentation"
             accesskey="P">previous</a> |
          <a href="minimap.html" title="Minimap2 Module Documentation"
             accesskey="N">next</a> |
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="module-ensembl.tools.anno.protein_annotation.genblast">
<span id="genblast-module-documentation"></span><h1>Genblast Module Documentation<a class="headerlink" href="#module-ensembl.tools.anno.protein_annotation.genblast" title="Link to this heading">¶</a></h1>
<p>GenBlast identifies homologous gene sequences in genomic databases.
One of the key features of GenBlast is its flexibility to handle
comparative genomics tasks and accurately identify homologs even when
the sequences have undergone significant evolutionary changes.
This capability makes it a valuable resource for researchers studying gene
evolution, gene families, and gene function across diverse species.
GenBlast has been widely used in various genomic analyses and is available as
a standalone command-line tool or as part of different bioinformatics pipelines.
Researchers in the field of comparative genomics and gene function analysis
often rely on GenBlast to perform sensitive homology searches and obtain
valuable insights into the evolutionary relationships and functional conservation
of genes in different organisms.</p>
<p>She, R., Chu, J.S., Uyar, B., Wang, J., Wang, K., and Chen, N. (2011).
GenBlastA: enabling BLAST to identify homologous gene sequences.
Genome Res., 21(5): 936-949.</p>
<dl class="py function">
<dt class="sig sig-object py" id="ensembl.tools.anno.protein_annotation.genblast.run_genblast">
<span class="sig-prename descclassname"><span class="pre">ensembl.tools.anno.protein_annotation.genblast.</span></span><span class="sig-name descname"><span class="pre">run_genblast</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">masked_genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_dataset</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_intron_length</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genblast_timeout_secs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10800</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genblast_bin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">PosixPath('genblast')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">convert2blastmask_bin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">PosixPath('convert2blastmask')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">makeblastdb_bin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">PosixPath('makeblastdb')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['uniprot',</span> <span class="pre">'orthodb']</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/ensembl/tools/anno/protein_annotation/genblast.html#run_genblast"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ensembl.tools.anno.protein_annotation.genblast.run_genblast" title="Link to this definition">¶</a></dt>
<dd><p>Executes GenBlast on genomic slices</p>
<blockquote>
<div><dl class="field-list simple">
<dt class="field-odd">param masked_genome<span class="colon">:</span></dt>
<dd class="field-odd"><p>Masked genome file path.</p>
</dd>
<dt class="field-even">type masked_genome<span class="colon">:</span></dt>
<dd class="field-even"><p>Path</p>
</dd>
<dt class="field-odd">param output_dir<span class="colon">:</span></dt>
<dd class="field-odd"><p>Working directory path.</p>
</dd>
<dt class="field-even">type output_dir<span class="colon">:</span></dt>
<dd class="field-even"><p>Path</p>
</dd>
<dt class="field-odd">param protein_dataset<span class="colon">:</span></dt>
<dd class="field-odd"><p>Protein dataset (Uniprot/OrthoDb) path.</p>
</dd>
<dt class="field-even">type protein_dataset<span class="colon">:</span></dt>
<dd class="field-even"><p>Path</p>
</dd>
<dt class="field-odd">param genblast_timeout_secs<span class="colon">:</span></dt>
<dd class="field-odd"><p>Time for timeout (sec).</p>
</dd>
<dt class="field-even">type genblast_timeout_secs<span class="colon">:</span></dt>
<dd class="field-even"><p>int, default 10800</p>
</dd>
<dt class="field-odd">param max_intron_length<span class="colon">:</span></dt>
<dd class="field-odd"><p>Maximum intron length.</p>
</dd>
<dt class="field-even">type max_intron_length<span class="colon">:</span></dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">param genblast_bin<span class="colon">:</span></dt>
<dd class="field-odd"><p>Software path.</p>
</dd>
<dt class="field-even">type genblast_bin<span class="colon">:</span></dt>
<dd class="field-even"><p>Path, default genblast</p>
</dd>
<dt class="field-odd">param convert2blastmask_bin<span class="colon">:</span></dt>
<dd class="field-odd"><p>Software path.</p>
</dd>
<dt class="field-even">type convert2blastmask_bin<span class="colon">:</span></dt>
<dd class="field-even"><p>Path, default convert2blastmask</p>
</dd>
<dt class="field-odd">param makeblastdb_bin<span class="colon">:</span></dt>
<dd class="field-odd"><p>Software path.</p>
</dd>
<dt class="field-even">type makeblastdb_bin<span class="colon">:</span></dt>
<dd class="field-even"><p>Path, default makeblastdb</p>
</dd>
<dt class="field-odd">param genblast_timeout<span class="colon">:</span></dt>
<dd class="field-odd"><p>seconds</p>
</dd>
<dt class="field-even">type genblast_timeout<span class="colon">:</span></dt>
<dd class="field-even"><p>int, default 1</p>
</dd>
<dt class="field-odd">param num_threads<span class="colon">:</span></dt>
<dd class="field-odd"><p>int, number of threads.</p>
</dd>
<dt class="field-even">type num_threads<span class="colon">:</span></dt>
<dd class="field-even"><p>int, default 1</p>
</dd>
<dt class="field-odd">param protein_set<span class="colon">:</span></dt>
<dd class="field-odd"><p>Source</p>
</dd>
<dt class="field-even">type str<span class="colon">:</span></dt>
<dd class="field-even"><p>[“uniprot”, “orthodb”]</p>
</dd>
<dt class="field-odd">return<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          
          <h3>Table of Contents</h3>
          <p class="caption" role="heading"><span class="caption-text">Index</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">API Setup and installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="cpg.html">CpG Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="dust.html">DustMasker Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="eponine.html">Eponine Module Documentation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Genblast Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="minimap.html">Minimap2 Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="red.html">Red Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="repeatmasker.html">Repeatmasker Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="scallop.html">Scallop Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="star.html">STAR Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="stringtie.html">Stringtie Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="trf.html">TRF Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="trnascan.html">tRNAscan-SE Module Documentation</a></li>
</ul>

          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
            </form>
          </div>

        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="eponine.html" title="Eponine Module Documentation"
              >previous</a> |
            <a href="minimap.html" title="Minimap2 Module Documentation"
              >next</a> |
            <a href="py-modindex.html" title="Python Module Index"
              >modules</a> |
            <a href="genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
              <br/>
              <a href="_sources/genblast.rst.txt"
                rel="nofollow">Show Source</a>
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
    &#169; Copyright [2016-2023] EMBL-European Bioinformatics Institute.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.5.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>