<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ensembl.tools.anno.snc_rna_annotation.trnascan &#8212; ensembl-anno 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../../../_static/agogo.css?v=cc331ede" />
    <script data-url_root="../../../../../" id="documentation_options" src="../../../../../_static/documentation_options.js?v=e031e9a9"></script>
    <script src="../../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <link rel="canonical" href="https://ensembl.github.io/ensembl-anno/_modules/ensembl/tools/anno/snc_rna_annotation/trnascan.html" />
    <link rel="index" title="Index" href="../../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../../search.html" /> 
  </head><body>
    <div class="header-wrapper" role="banner">
      <div class="header">
        <div class="headertitle"><a
          href="../../../../../index.html">ensembl-anno 0.1 documentation</a></div>
        <div class="rel" role="navigation" aria-label="related navigation">
          <a href="../../../../../py-modindex.html" title="Python Module Index"
             >modules</a> |
          <a href="../../../../../genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for ensembl.tools.anno.snc_rna_annotation.trnascan</h1><div class="highlight"><pre>
<span></span><span class="c1"># See the NOTICE file distributed with this work for additional information</span>
<span class="c1"># regarding copyright ownership.</span>
<span class="c1">#</span>
<span class="c1"># Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);</span>
<span class="c1"># you may not use this file except in compliance with the License.</span>
<span class="c1"># You may obtain a copy of the License at</span>
<span class="c1">#</span>
<span class="c1">#     http://www.apache.org/licenses/LICENSE-2.0</span>
<span class="c1">#</span>
<span class="c1"># Unless required by applicable law or agreed to in writing, software</span>
<span class="c1"># distributed under the License is distributed on an &quot;AS IS&quot; BASIS,</span>
<span class="c1"># WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.</span>
<span class="c1"># See the License for the specific language governing permissions and</span>
<span class="c1"># limitations under the License.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">tRNAscan-SE identifies 99-100% of transfer RNA genes in DNA sequence while</span>
<span class="sd">giving less than one false positive per 15 gigabases.</span>
<span class="sd">Lowe TM, Eddy SR: tRNAscan-SE: a program for improved detection of transfer</span>
<span class="sd">RNA genes in genomic sequence.</span>
<span class="sd">Nucleic Acids Res. 1997, 25(5):955-64. [PMID: 9023104]</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;run_trnascan&quot;</span><span class="p">]</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">logging.config</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="kn">from</span> <span class="nn">os</span> <span class="kn">import</span> <span class="n">PathLike</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span>

<span class="kn">from</span> <span class="nn">ensembl.tools.anno.utils._utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">check_exe</span><span class="p">,</span>
    <span class="n">check_file</span><span class="p">,</span>
    <span class="n">create_dir</span><span class="p">,</span>
    <span class="n">check_gtf_content</span><span class="p">,</span>
    <span class="n">get_seq_region_length</span><span class="p">,</span>
    <span class="n">get_slice_id</span><span class="p">,</span>
    <span class="n">slice_output_to_gtf</span><span class="p">,</span>
    <span class="n">get_sequence</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<div class="viewcode-block" id="run_trnascan"><a class="viewcode-back" href="../../../../../trnascan.html#ensembl.tools.anno.snc_rna_annotation.trnascan.run_trnascan">[docs]</a><span class="k">def</span> <span class="nf">run_trnascan</span><span class="p">(</span>
    <span class="n">genome_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">trnascan_bin</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;tRNAscan-SE&quot;</span><span class="p">),</span>
    <span class="n">trnascan_filter</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;EukHighConfidenceFilter&quot;</span><span class="p">),</span>
    <span class="n">num_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Executes tRNAscan-SE on genomic slices</span>
<span class="sd">        :param genome_file: Genome file path.</span>
<span class="sd">        :type genome_file: PathLike</span>
<span class="sd">        :param output_dir:  working directory path.</span>
<span class="sd">        :type output_dir: Path</span>
<span class="sd">        :param trnascan_bin: tRNAscan-SE software path.</span>
<span class="sd">        :type trnascan_bin: Path, default tRNAscan-SE</span>
<span class="sd">        :param trnascan_filter: tRNAscan-SE filter set path.</span>
<span class="sd">        :type trnascan_filter: Path, default EukHighConfidenceFilter</span>
<span class="sd">        :param num_threads: int, number of threads.</span>
<span class="sd">        :type num_threads: int, default 1</span>

<span class="sd">        :return: None</span>
<span class="sd">        :rtype: None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">check_exe</span><span class="p">(</span><span class="n">trnascan_bin</span><span class="p">)</span>
    <span class="n">check_file</span><span class="p">(</span><span class="n">trnascan_filter</span><span class="p">)</span>
    <span class="n">trnascan_dir</span> <span class="o">=</span> <span class="n">create_dir</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;trnascan_output&quot;</span><span class="p">)</span>
    <span class="n">output_file</span> <span class="o">=</span> <span class="n">trnascan_dir</span> <span class="o">/</span> <span class="s2">&quot;annotation.gtf&quot;</span>
    <span class="k">if</span> <span class="n">output_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">transcript_count</span> <span class="o">=</span> <span class="n">check_gtf_content</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="s2">&quot;transcript&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">transcript_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Trnascan gtf file exists, skipping analysis&quot;</span><span class="p">)</span>
            <span class="k">return</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating list of genomic slices&quot;</span><span class="p">)</span>
    <span class="n">seq_region_to_length</span> <span class="o">=</span> <span class="n">get_seq_region_length</span><span class="p">(</span><span class="n">genome_file</span><span class="p">,</span> <span class="mi">5000</span><span class="p">)</span>
    <span class="n">slice_ids_per_region</span> <span class="o">=</span> <span class="n">get_slice_id</span><span class="p">(</span><span class="n">seq_region_to_length</span><span class="p">,</span> <span class="mi">1000000</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">5000</span><span class="p">)</span>
    <span class="n">trnascan_cmd</span> <span class="o">=</span> <span class="p">[</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">trnascan_bin</span><span class="p">),</span>
        <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span>
        <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;-f&quot;</span><span class="p">,</span>
        <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;-H&quot;</span><span class="p">,</span>  <span class="c1"># show both primary and secondary structure components to covariance model bit scores</span>
        <span class="s2">&quot;-q&quot;</span><span class="p">,</span>  <span class="c1"># quiet mode</span>
        <span class="s2">&quot;--detail&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-Q&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running tRNAscan-SE&quot;</span><span class="p">)</span>
    <span class="n">pool</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">num_threads</span><span class="p">)</span>  <span class="c1"># pylint: disable=consider-using-with</span>
    <span class="k">for</span> <span class="n">slice_id</span> <span class="ow">in</span> <span class="n">slice_ids_per_region</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span>
            <span class="n">_multiprocess_trnascan</span><span class="p">,</span>
            <span class="n">args</span><span class="o">=</span><span class="p">(</span>
                <span class="n">trnascan_cmd</span><span class="p">,</span>
                <span class="n">slice_id</span><span class="p">,</span>
                <span class="n">genome_file</span><span class="p">,</span>
                <span class="n">trnascan_filter</span><span class="p">,</span>
                <span class="n">trnascan_dir</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
    <span class="n">slice_output_to_gtf</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="n">trnascan_dir</span><span class="p">,</span> <span class="n">unique_ids</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">file_extension</span><span class="o">=</span><span class="s2">&quot;.trna.gtf&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">gtf_file</span> <span class="ow">in</span> <span class="n">trnascan_dir</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;*.trna.gtf&quot;</span><span class="p">):</span>
        <span class="n">gtf_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span></div>


<span class="k">def</span> <span class="nf">_multiprocess_trnascan</span><span class="p">(</span>
    <span class="n">trnascan_cmd</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">slice_id</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">genome_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">trnascan_filter</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">trnascan_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run tRNAscan-SE on multiprocess on genomic slices</span>
<span class="sd">    Args:</span>
<span class="sd">        trnascan_cmd: tRNAscan-SE command to execute.</span>
<span class="sd">        slice_id: Slice Id to run tRNAscan-SE on.</span>
<span class="sd">        genome_file : Genome file.</span>
<span class="sd">        trnascan_dir : tRNAscan-SE output dir.</span>
<span class="sd">        trnascan_filter: tRNAscan-SE filter set.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">region_name</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">slice_id</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Processing slice to find tRNAs using tRNAscan-SE:</span><span class="si">%s</span><span class="s2">:</span><span class="si">%s</span><span class="s2">:</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">region_name</span><span class="p">,</span>
        <span class="n">start</span><span class="p">,</span>
        <span class="n">end</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">get_sequence</span><span class="p">(</span><span class="n">region_name</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">),</span> <span class="mi">1</span><span class="p">,</span> <span class="n">genome_file</span><span class="p">,</span> <span class="n">trnascan_dir</span><span class="p">)</span>
    <span class="n">slice_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">region_name</span><span class="si">}</span><span class="s2">.rs</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s2">.re</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">slice_file</span> <span class="o">=</span> <span class="n">trnascan_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">slice_name</span><span class="si">}</span><span class="s2">.fa&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">slice_file</span><span class="p">,</span> <span class="s2">&quot;w+&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">region_out</span><span class="p">:</span>
        <span class="n">region_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&gt;</span><span class="si">{</span><span class="n">region_name</span><span class="si">}</span><span class="se">\n</span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># trnscan output</span>
    <span class="n">region_results</span> <span class="o">=</span> <span class="n">trnascan_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">slice_name</span><span class="si">}</span><span class="s2">.trna.gtf&quot;</span>
    <span class="n">output_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">slice_file</span><span class="si">}</span><span class="s2">.trna&quot;</span><span class="p">)</span>
    <span class="n">ss_output_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">.ss&quot;</span><span class="p">)</span>
    <span class="c1"># filtering</span>
    <span class="n">filter_prefix_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">slice_name</span><span class="si">}</span><span class="s2">.filt&quot;</span>
    <span class="n">filter_output_file</span> <span class="o">=</span> <span class="n">trnascan_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">filter_prefix_file</span><span class="si">}</span><span class="s2">.out&quot;</span>
    <span class="n">filter_log_file</span> <span class="o">=</span> <span class="n">trnascan_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">filter_prefix_file</span><span class="si">}</span><span class="s2">.log&quot;</span>
    <span class="n">filter_ss_file</span> <span class="o">=</span> <span class="n">trnascan_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">filter_prefix_file</span><span class="si">}</span><span class="s2">.ss&quot;</span>
    <span class="c1"># trnascan_cmd = generic_trnascan_cmd.copy()</span>
    <span class="n">trnascan_cmd</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">trnascan_cmd</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">trnascan_cmd</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">slice_file</span><span class="p">),</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">ss_output_file</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;tRNAscan-SE command: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">trnascan_cmd</span><span class="p">))</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">trnascan_cmd</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># If the trnascan output is empty there is no need to go on with filtering</span>
    <span class="k">if</span> <span class="n">output_file</span><span class="o">.</span><span class="n">stat</span><span class="p">()</span><span class="o">.</span><span class="n">st_size</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">output_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
        <span class="n">slice_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>
        <span class="n">ss_output_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="n">filter_cmd</span> <span class="o">=</span> <span class="p">[</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">trnascan_filter</span><span class="p">),</span>
        <span class="s2">&quot;--result&quot;</span><span class="p">,</span>  <span class="c1"># tRNAscan-SE output file used as input</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span>
        <span class="s2">&quot;--ss&quot;</span><span class="p">,</span>  <span class="c1"># tRNAscan-SE secondary structure file used as input</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">ss_output_file</span><span class="p">),</span>
        <span class="s2">&quot;--output&quot;</span><span class="p">,</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">trnascan_dir</span><span class="p">),</span>
        <span class="s2">&quot;--prefix&quot;</span><span class="p">,</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">filter_prefix_file</span><span class="p">),</span>
    <span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;tRNAscan-SE filter command: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">filter_cmd</span><span class="p">))</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">filter_cmd</span><span class="p">)</span>  <span class="c1"># pylint:disable=subprocess-run-check</span>
    <span class="n">_create_trnascan_gtf</span><span class="p">(</span><span class="n">region_results</span><span class="p">,</span> <span class="n">filter_output_file</span><span class="p">,</span> <span class="n">region_name</span><span class="p">)</span>
    <span class="n">output_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">slice_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ss_output_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">Path</span><span class="p">(</span><span class="n">filter_prefix_file</span><span class="p">)</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">filter_log_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">filter_ss_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">filter_output_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_create_trnascan_gtf</span><span class="p">(</span><span class="n">region_results</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">filter_output_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">region_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read the fasta file and save the content in gtf format</span>
<span class="sd">    All the genomic slices are collected in a single gtf output</span>
<span class="sd">    Args:</span>
<span class="sd">        region_results : GTF file with the results per region.</span>
<span class="sd">        filter_file : GTF file with the filtered results per region.</span>
<span class="sd">        region_name :Coordinates of genomic slice.</span>

<span class="sd">    tRNAscan-SE output format:</span>
<span class="sd">    col0: GtRNAdb Gene Symbol - gene ID in corresponding genome</span>
<span class="sd">    col1: tRNAscan-SE ID - tRNA ID in tRNAscan-SE prediction results</span>
<span class="sd">    col2-3: Locus - Genomic coordinates of predicted gene</span>
<span class="sd">    col4: Isotype (from Anticodon) - tRNA isotype determined by anticodon</span>
<span class="sd">    col5: Anticodon - anticodon of predicted tRNA gene</span>
<span class="sd">    col6-7: Intron boundaries</span>
<span class="sd">    col8: General tRNA Model Score - covariance model bit score from tRNAscan-SE results</span>
<span class="sd">    col9: Best Isotype Model - best matching (highest scoring) isotype determined</span>
<span class="sd">    by isotype-specific covariance model classification</span>
<span class="sd">    col10-11-12: Anticodon and Isotype Model Agreement - consistency between anticodon</span>
<span class="sd">    from predicted gene sequence and best isotype model</span>
<span class="sd">    col13: Features - special gene features that may include gene set categorization,</span>
<span class="sd">    number of introns, possible pseudogenes, possible truncation, or base-pair mismatches</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filter_output_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">trna_in</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span>
        <span class="n">region_results</span><span class="p">,</span> <span class="s2">&quot;w+&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">trna_out</span><span class="p">:</span>
        <span class="n">gene_counter</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">trna_in</span><span class="p">:</span>
            <span class="n">result_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^&quot;</span> <span class="o">+</span> <span class="n">region_name</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">result_match</span><span class="p">:</span>
                <span class="n">results</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;+&quot;</span>
                <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
                    <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>
                    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">end</span><span class="p">,</span> <span class="n">start</span>
                <span class="n">biotype</span> <span class="o">=</span> <span class="s2">&quot;tRNA&quot;</span> <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;high confidence set&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;tRNA_pseudogene&quot;</span>
                <span class="n">transcript_string</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">region_name</span><span class="si">}</span><span class="se">\t</span><span class="s2">tRNAscan</span><span class="se">\t</span><span class="s2">transcript</span><span class="se">\t</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="se">\t</span><span class="s2">.</span><span class="se">\t</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">strand</span><span class="si">}</span><span class="se">\t</span><span class="s1">.</span><span class="se">\t</span><span class="s1">gene_id &quot;</span><span class="si">{</span><span class="n">gene_counter</span><span class="si">}</span><span class="s1">&quot;; transcript_id &#39;</span>
                    <span class="sa">f</span><span class="s1">&#39;&quot;</span><span class="si">{</span><span class="n">gene_counter</span><span class="si">}</span><span class="s1">&quot;; biotype &quot;</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s1">&quot;;</span><span class="se">\n</span><span class="s1">&#39;</span>
                <span class="p">)</span>
                <span class="n">exon_string</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">region_name</span><span class="si">}</span><span class="se">\t</span><span class="s2">tRNAscan</span><span class="se">\t</span><span class="s2">exon</span><span class="se">\t</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="se">\t</span><span class="s2">.</span><span class="se">\t</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">strand</span><span class="si">}</span><span class="se">\t</span><span class="s1">.</span><span class="se">\t</span><span class="s1">gene_id &quot;</span><span class="si">{</span><span class="n">gene_counter</span><span class="si">}</span><span class="s1">&quot;; transcript_id &#39;</span>
                    <span class="sa">f</span><span class="s1">&#39;&quot;</span><span class="si">{</span><span class="n">gene_counter</span><span class="si">}</span><span class="s1">&quot;; exon_number &quot;1&quot;; biotype &quot;</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s1">&quot;;</span><span class="se">\n</span><span class="s1">&#39;</span>
                <span class="p">)</span>
                <span class="n">trna_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">transcript_string</span><span class="p">)</span>
                <span class="n">trna_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">exon_string</span><span class="p">)</span>
                <span class="n">trna_out</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
                <span class="n">gene_counter</span> <span class="o">+=</span> <span class="mi">1</span>


<span class="k">def</span> <span class="nf">parse_args</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse command line arguments.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s2">&quot;tRNAscan-SE&#39;s arguments&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--genome_file&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Genome file path&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--trnascan_bin&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;tRNAscan-SE&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;tRNAscan-SE executable path&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--trnascan_filter&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;/hps/software/users/ensembl/ensw/C8-MAR21-sandybridge/linuxbrew/bin/EukHighConfidenceFilter&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;tRNAscan-SE filter path&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--output_dir&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output directory path&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--num_threads&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of threads&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;tRNAscan-SE&#39;s entry-point.&quot;&quot;&quot;</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parse_args</span><span class="p">()</span>

    <span class="n">log_file_path</span> <span class="o">=</span> <span class="n">create_dir</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;log&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;trnascan.log&quot;</span>
    <span class="n">loginipath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span><span class="o">.</span><span class="n">parents</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="o">/</span> <span class="s2">&quot;conf&quot;</span> <span class="o">/</span> <span class="s2">&quot;logging.conf&quot;</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">fileConfig</span><span class="p">(</span>
        <span class="n">loginipath</span><span class="p">,</span>
        <span class="n">defaults</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;logfilename&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">)},</span>
        <span class="n">disable_existing_loggers</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">run_trnascan</span><span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">genome_file</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span>
        <span class="n">args</span><span class="o">.</span><span class="n">trnascan_bin</span><span class="p">,</span>
        <span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">trnascan_filter</span><span class="p">),</span>
        <span class="n">args</span><span class="o">.</span><span class="n">num_threads</span><span class="p">,</span>
    <span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          
          <h3>Table of Contents</h3>
          <p class="caption" role="heading"><span class="caption-text">Index</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../../install.html">API Setup and installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../cmsearch.html">cmsearch Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../cpg.html">CpG Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../dust.html">DustMasker Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../eponine.html">Eponine Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../genblast.html">Genblast Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../minimap.html">Minimap2 Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../red.html">Red Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../repeatmasker.html">Repeatmasker Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../scallop.html">Scallop Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../star.html">STAR Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../stringtie.html">Stringtie Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../trf.html">TRF Module Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../../trnascan.html">tRNAscan-SE Module Documentation</a></li>
</ul>

          <div role="search">
            <h3 style="margin-top: 1.5em;">Search</h3>
            <form class="search" action="../../../../../search.html" method="get">
                <input type="text" name="q" />
                <input type="submit" value="Go" />
            </form>
          </div>

        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <div role="navigation" aria-label="related navigaton">
            <a href="../../../../../py-modindex.html" title="Python Module Index"
              >modules</a> |
            <a href="../../../../../genindex.html" title="General Index"
              >index</a>
          </div>
          <div role="note" aria-label="source link">
          </div>
        </div>

        <div class="right">
          
    <div class="footer" role="contentinfo">
    &#169; Copyright [2016-2024] EMBL-European Bioinformatics Institute.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.1.2.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>